Readme (#6)
marmelab/liMonade: Readme.md

Traversable allow you to do just that.
Go to the List section for an example with a List of Tasks becoming a Task of a List.

The sequence method just permute the two monads:
```ts
const id = Identity.of(List([1, 2, 3])); // Identity<List<[1, 2, 3]>>
id.sequence(Maybe.of); // List<[Identity<1>, Identity<2>, Identity<3>]>
```

The traverse method first transforms the value to an applicative, then do the same as sequence.

```js
const stringToList = v => List(v.split(''));
const id = Identity.of('hey')); // Identity<'hey'>
id.traverse(stringToList, List.of); // List<[Identity<h>, Identity<e>, Identity<y>]>
```

See List for a more practical example with a List of asynchronous Task becoming a Task of a list of value.

## Identity can lift

Finally Identity offer a lift function.
lift is an helper that take a function and returns a new function that wraps its result in an Identity.

With lift when you want a function that returns an identity, you can take a normal function then lift it.

```js
const double = v => v * 2;
doubleThatReturnsAnIdentity = Identity.lift(double);

doubleThatReturnsAnIdentity(5); // Identity<10>
```

# Maybe

> Maybe is used to hanlde null value. That is, if we have no value, then it won't call the function and so it will keep a null value. No more need to test if we have a value or not.

Reviewer: s/hanlde/handle/

It is useful to handle computations that depends on an external factor. Like user input for example.

### Example

Let's say that we have an input field and an output where we want to display what the user typed, but in uppercase

```html
<input id="input" onInput="displayInput()" />

<p id="output"></p>
```

A side note for previous Monad, the of method was the same as the constructor for other Monad. FOr IO that is not the xcase anymore.
IO take the side effect function that will return the value.
IO.of take a value and wrap it in a function, creating a side effect from it. Sort of.

```js
const inputIO = IO(() => document.getElementById('input').value);
const outputIO = IO.of(
    value => document.getElementById('output').innerText = value
); // notice how outputIO hold a function and not a value
const toUpperCase = v => v.toUpperCase()

const displayInput = outputIO.ap(
    inputIO.map(toUpperCase)
).execute;
```

> // TODO add a codepen link once the library is published

Reviewer: TODO found

## Maybe api

> - Maybe take a value and return a maybe, if the passed value is null or undefined, all operation on the maybe to change the value will be ignored.

Reviewer 2: Maybe takes a value and returns a maybe. When the value given in argument is null or undefined, all the operations applied to the maybe monad to change its value will be ignored.

- Maybe.of: same as Maybe

> - Maybe.lift: Takes a function and wrap its return value in Maybe

Reviewer 2: s/wrap/wraps/

```ts
// Maybe.lift(fn: A => B): (v: A) => Maybe<B>;
const fn = Maybe.lift(v => v * 2);
// is the same as
const fn = v => Maybe.of(v * 2);
```

> - maybe.map: Take a function and apply it to the value if there is one

Reviewer 2: takes and applies

```ts
// Maybe<A>.map<fn: A => B>: Maybe<B>;
Maybe.of(5).map(v => v * 2);
-> Maybe<10>;
// Maybe<null | undefined>.map(fn: A => B): Maybe<null | undefined>;
Maybe.of(null).map(v => v * 2);
-> Maybe<null>;
```

> - maybe.flatten: If the value is another Maybe merge the two together. If there is no value, do nothing.

Reviewer 2: merge the current maybe with the one given in argument if there is one. Otherwise, does nothing.

```ts
// Maybe<Maybe<A>>.flatten(): Maybe<A>;
Maybe.of(Maybe.of(5)).flatten();
-> Maybe<5>;
// Maybe<null | undefined>.flatten(): Maybe<null | undefined>;
Maybe.of(null).flatten();
-> Maybe<null>;
```
- maybe.chain
```ts
//Maybe<A>.chain(fn: A => Maybe<B>): Maybe<B>;
Maybe.of(5).chain(v => Maybe.of(v * 2));
-> Maybe<10>;
// Maybe<nothing>.chain(fn: A => Maybe<B>): Maybe<nothing>;
Maybe.of(null).chain(v => Maybe.of(v * 2));
-> Maybe<null>;
```
- maybe.ap
```ts
// Maybe<A => B>.ap(other: Maybe<A>): Maybe<B>;
Maybe.of(v => v * 2).ap(Maybe.of(5));
-> Maybe<10>;
// Maybe<nothing>.ap(other: Maybe<A>): Maybe<nothing>;
Maybe.of(null).ap(Maybe.of(5));
-> Maybe<null>;
// Maybe<A => B>.ap(other: Maybe<nothing>): Maybe<nothing>;
Maybe.of(v => v * 2).ap(Maybe.of(null));
-> Maybe<null>;
```

> - maybe.sequence: when the maybe hold an applicative, it will swap the maybe with the applicative. If the maybe hold nothing it will place the maybe inside an apllicative.

Reviewer 2: - maybe.sequence: when the maybe holds an applicative, it will swap the maybe with the applicative. When the maybe does not hold anything, it will place the maybe inside an applicative.

```ts
/**
 * Maybe<Applicative<A>>.sequence(
 *     of: Value => Applicative<Value>,
 * ): Applicative<Maybe<A>>;
 * */
Maybe.of(Identity.of(5)).sequence(Identity.of);
    -> Identity<Maybe<5>>;
/*
 * Maybe<nothing>.sequence<A>(
 *     of: Value => Applicative<Value>,
 * ): Applicative<Maybe<nothing>>;
 * */
Maybe.of(null).sequence(Identity.of);
    -> Identity<Maybe<null>>;
```
- maybe.traverse: map a function returning an applicative and then swap the applicative with the maybe.

> If the maybe hold nothing, an Apllicative<Maybe<nothing>> will be returned

Reviewer 2: ```suggestion

 When the maybe holds nothing, an Applicative<Maybe<nothing>> is returned
```

```ts
/*
 * Maybe<Applicative<A>>.traverse(
 *     fn: A => Applicative<B>,
 *     of: Value => Applicative<Value>,
 * ): Applicative<Maybe<B>>;
 * */
Maybe.of(5).traverse(v => Identity.of(v * 2));
    -> Identity<Maybe<10>>;
/*
 * Maybe<nothing>.traverse(
 *     of: Value => Applicative<Value>,
 *     fn: A => Applicative<B>,
 * ): Applicative<Maybe<nothing>>;
 * */
Maybe.of(null).traverse(v => Identity.of(v * 2));
    -> Identity<Maybe<10>>;
```

> - maybe.isNothing: return true if the maybe holds nothing (null or undefined)

Reviewer 2: returns

```ts
Maybe<A>.isNothing(): false;
Maybe<nothing>.isNothing(): true;
```

- maybe.isJust: return true if the maybe holds a value

```ts
Maybe<A>.isJust(): true;
Maybe<nothing>.isJust(): false;
```

> - maybe.getOrElse: return the Maybe value or the given value if the maybe hold nothing.

Reviewer 2: returns/holds

```ts
Maybe<A>.getOrElse(defaultValue: B): A;
Maybe<nothing>.getOrElse(defaultValue: B): B;
```

# Either

> Either is used to handle possible error, if receiving an error it will preserve it, otherwise it will map normally. If any mapped function would trhow an error, the error will be placed in the Either.

Reviewer 2: possible errors.
what does map normally means: do you mean its behavior is like a classical map function ?
trhow-> throw

- Either.of: same as Either
- Either.Right: same as Either but accept only non error value. Return a Right Either
- Either.Left: same as Either but accept only error value. Return a Left Either

> - Either.lift: Take a function and wrap its return value in a Right Either or its thrown error in a left Either

Reviewer 2: +s
its thrown ?

```ts
// Either.lift(fn: A => B): (v: A) => Either<B>;
Either.lift(v => v * 2);
    -> v => Either.of(v * 2);
// Either.lift(fn: A => throw Error): (v: A) => Either<Error>;
Either.lift(v => throw new Error('Boom'));
    -> v => Either.of(new Error('Boom'));
```

- either.map: Take a function and apply it to the value if it's not an error. If the function throw an error a Left Either of the error is returned
```ts
// Either<A>.map(fn: A => B): Either<B>;
Either.of(5).map(v => v * 2);
-> Either<10>;

// Either<Error>.map(fn: A => B): Either<Error>;
Either.of(new Error('Boom')).map(v => v * 2);
-> Either<new Error('Boom')>;
```

> - either.flatten: If the value is another Either merge the two together. If the value is an Error do nothing.

Reviewer 2: if (no capital)

Either.of(5).chain(v => v.concat('uh'));
    -> Either<Error('v.concat is not a function')>
```
- either.ap
```ts
// Either<A => B>.ap(other: Either<A>): Either<B>;
Either.of(v => v * 2).ap(Either.of(5));
    -> Either<10>
// Either<Error>.ap(other: Either<A>): Either<Error>;
Either.of(new Error('Boom')).ap(Either.of(5));
    -> Either<new Error('Boom')>
// Either<A => B>.ap(other: Either<Error>): Either<Error>;
Either.of(v => v * 2).ap(Either.of(new Error('Boom')));
    -> Either<new Error('Boom')>
```
- either.sequence: when the either hold an applicative, it will swap the either with the applicative. If the either hold an error it will place the either inside an apllicative.
```ts
/*
 * Either<Applicative<A>>.sequence(
 *     of: Value => Applicative<Value>,
 * ): Applicative<Either<A>>;
 * */
Either.of(Identity.of(5)).sequence(Identity.of);
    -> Identity<Either<5>>
/*
 * Either<Error>.sequence(
 *     of: Value => Applicative<Value>,
 * ): Applicative<Either<Error>>;
 * */
Either.of(new Error('Boom')).sequence(Identity.of);
    -> Identity<Either<new Error('Boom')>>
```
- maybe.traverse: map a function returning an applicative and then swap the applicative with the either.

> If the either hold an error, an Apllicative<Maybe<Error>> will be returned

Reviewer 2: Applicative

```ts
/*
 * Either<Applicative<A>>.traverse(
 *     fn: A => Applicative<B>,
 *     of: Value => Applicative<Value>,
 * ): Applicative<Either<B>>;
 * */
Either.of(5).traverse(v => Identity.of(v * 2));
    -> Identity<Either<10>>
/*
 * Either<Error>.sequence(
 *     fn: A => Applicative<B>,
 *     of: Value => Applicative<Value>,
 * ): Applicative<Either<Error>>;
 * */
Either.of(5).traverse(v => throw new Error('Boom'));
    -> Identity<Either<new Error('Boom')>>
```

> - maybe.catch: works like map, it accept a function, and will call it if the value is an error. It Will transform the Error in a value an return a Right Error.

Reviewer 2: +s
and returns

```ts
// Either<A>.catch(fn: Error => B): Either<A>;
Either.of(5).catch(error => error.message);
    -> Either<5>
// Either<Error>.catch(fn: Error => B): Either<B>;
Either.of(new Error('Boom')).catch(error => error.message);
    -> Either<'Boom'>
```
- maybe.isLeft: return true if the maybe holds an error
```ts
Either<A>.isLeft(): false;
Either<Error>.isLeft(): true;
```
- maybe.isRight: return true if the maybe holds a value
```ts
Either<A>.isRight(): true;
Either<Error>.isRight(): false;
```
- maybe.get: return the value or throw the error
```ts
Either<A>.get(): A;
Either<Error>.get(): throw Error;
```

## IO

> IO is a shorthand for `Input/Output` IO is used to handle side effect. Instead of holding the value directly, we hold a function that will return the value. The important distinction compared to the previous monads is that IO is lazy. We can map all we want, nothing will happen until you call its execute method.

Reviewer 2: side effects
IO holds
lazy: the map function can be called several times, nothing will be executed until the `execute` method is called

> It is useful to handle computation that depends on an external factor. Like user input.

Reviewer 2: computations
like use input for example

getJoke(9999).then(console.log, console.log); // "joke not found"
```

### Task api

- Task takes an asynchronous function and return a Task

```ts
/*
 * Task(
 *     cps: (resolve: Value => void, reject: Error => void) => void,
 * ): Task<Value>
 * */
Task((resolve, reject) => resolve('success'));
    -> Task<'success'>
```

- Task.of: take a value and return a Task that will allow to operate on the value

```ts
// Task.of<Value>(value: Value): Task<Value>
Task.of(5);
    -> Task<5>
```

> - Task.reject: take a value and return a Task rejected with this value. A rejected Task like a left either will ignore all operation on it except catch.

Reviewer 2: +s
operations
